ARCHITECTURE.md

```diff
@@ -0,0 +1,152 @@
+![BluepillIcon](doc/img/bluepill_text.png)
+
+![](https://github.com/linkedin/bluepill/workflows/master/badge.svg)
+
+## iOS Simulator Architecture
+
+- App (builds with) -> (Xcode/Bazel) -> `X.app` (packaged into an IPA, an archive of the app or "Bundle")
+- iPhones have ARM processors while Macs have `x86_64` (pre M1 at least)
+- Compilation results in an architecture specific binary (or app binary)
+
+## Bluepill Build How-To
+- `bluepill.sh` builds Bluepill in CI (It has a sample app)
+- Technically the script builds Bluepill & the sample app, and runs the tests
+- `BPSampleApp.app` is the "bundle", an x86_64 bundle but iPhone needs an ARM executable
+	- "Fat binary" will have both
+- Side note: iOS is a variant of MacOS and Android ships a VM with a Linux based Android kernel
+- Originally Apple Emulators translated ARM -> `x86_64` but that was very slow!
+	- Helped test iOS apps in Mac by giving you an `x86_64` executable
+- Now they compile all frameworks for `x86_64` and give you a sandbox in which you can build a native binary
+	- It's faster to get an exectuable to test this way
+
+
+## Some notes on iOS Simulators
+- `xcrun simctl boot` will boot the simulator (Alternatively run Simulator.App in /Applications/Xcode.App)
+- `ps axuw | grep 'launchd_sim'` -> This is the `launchd` process running inside the Simulator
+	- Since the Simulator isn't a VM (like Android) and shares the same Kernel as the underlying Mac OS, it's possible to inspect the processes running inside the simulator
+	- `ps axiw | grep 'CoreSimulator'`
+- `xcrun simctl install .... BPSampleApp.app`
+- Kernel runs with special persmissions (Ring 0) that user programs aren't privy to
+	- Copy of iOS runs in user space of MacOS Kernel
+- The VM model: Multiple Kernels that share the same processor
+	- CPU -> Kernel1 (OS1)/Kernel2 (OS2)/Kernel3 (OS2)
+- `otool -L MacOS/Simulator`
+	- Inspect Frameworks -> /Library/Developer/PrivateFrameworks/CoreSimulator..
+	- Bluepill links with PrivateFrameworks (like simctl)
+- All functionality is provided as a framework. Simluator UI uses this library and so does Bluepill
+- file CoreSimulator
+	- Shows you a binary with 2 architectures for M1 compatibility
+	- CoreSimulator.framework/Versions/A
+		- Poke around to use the framework directly in Bluepill
+
+## iOS Apps Test Classes
```

Comment:
Maybe we can just refer to some articles that explains the difference between XCTest and XCUITest